qgis/examples/elevation_tile_downloader/gui/tile_downloader_dlg.py

```diff
@@ -0,0 +1,148 @@
+    grid_bboxes = []
+
+    # loop through x and y range and create the grid
+    min_x, min_y = [floor(x) for x in (bounds.xMinimum(), bounds.yMinimum())]
+    max_x, max_y = [
+        ceil(x)
+        for x in (
+            bounds.xMinimum() + bounds.width(),
+            bounds.yMinimum() + bounds.height(),
+        )
+    ]
+    for x in range(min_x, max_x):
+        for y in range(min_y, max_y):
+            grid_bboxes.append(
+                QgsGeometry.fromRect(QgsRectangle(x, y, x + 1, y + 1))
+            )
+
+    return grid_bboxes
+
+
+class ElevationTileDownloaderDialog(
+    QDialog, Ui_ElevationTileDownloaderDialogBase
+):
+    def __init__(self, iface: QgisInterface, parent=None):
+        """Constructor."""
+        super(ElevationTileDownloaderDialog, self).__init__(parent)
+        self.iface = iface
+        self.setupUi(self)
+
+        self.progress_bar: Optional[QProgressBar] = None
+        self.task: Optional[QgsTask] = None
+        self.total = 0
+
+        # we only allow polygon layers
+        self.layer_choice.setFilters(QgsMapLayerProxyModel.PolygonLayer)
+        self.accepted.connect(self.download_tiles)
+
+        # change the text of the "Ok" button
+        self.button_box.button(QDialogButtonBox.Ok).setText("Download")
+
+    def download_tiles(self) -> None:
+        """
+        The plugin's main method; constructs the grid from selected input features,
+        creates and starts the download task and connects the plugin to task signals.
+        """
+        poly_layer: QgsVectorLayer = self.layer_choice.currentLayer()
+        out_dir = Path(self.output_dir.filePath())
+
+        # Collect the tile bounding boxes covering the input polygons
+        grid = set()
+        for feature in poly_layer.getFeatures():
+            # transform the feature's bounding box and create a 1 x 1 degree tile grid from it
+            feature_geom = to_wgs84(feature.geometry(), poly_layer.crs())
+            grid_bboxes = create_grid_from_bounds(feature_geom.boundingBox())
+            for grid_bbox in grid_bboxes:
+                if feature_geom.intersects(grid_bbox):
+                    print("bbox: ", grid_bbox)
+                    grid.add(grid_bbox)
+
+        self.total = len(grid)
+        if not self.total:
+            self.iface.messageBar().pushMessage(
+                "No tiles",
+                f"Layer {poly_layer.name()} does not intersect any tiles",
+            )
+            return
+
+        # Create a progress bar in the QGIS message bar
+        self.progress_bar = QProgressBar()
+        self.progress_bar.setMaximum(100)
+        self.progress_bar.setAlignment(Qt.AlignLeft | Qt.AlignVCenter)
+        progress_msg: QgsMessageBarItem = (
+            self.iface.messageBar().createMessage("Download Progress: ")
+        )
+        progress_msg.layout().addWidget(self.progress_bar)
+        self.iface.messageBar().pushWidget(progress_msg, Qgis.Info)
```

Comment:
this is basically the biggest change, the rest is just cosmetics. this will set up the download progress in the iface message bar of the main thread, I find that a better place than the task bar